rust-lib/src/lib.rs

```diff
@@ -214,38 +249,52 @@ pub extern "C" fn issue(runtime: &COpaqueStruct, json: *mut c_char) -> CResult {
     _issue(runtime, json).into()
 }
 
-#[derive(Debug, Deserialize)]
-struct TransferArgs {
-    inputs: Vec<OutPoint>,
-    allocate: Vec<Outcoins>,
-    #[serde(with = "serde_with::rust::display_fromstr")]
-    invoice: Invoice,
-    prototype_psbt: String,
-    consignment_file: String,
-    transaction_file: String,
-}
-
-fn _transfer(runtime: &COpaqueStruct, json: *mut c_char) -> Result<(), String> {
+fn _transfer(runtime: &COpaqueStruct, inputs: *mut c_char, allocate: *mut c_char,
+             invoice: *mut c_char, prototype_psbt: *mut c_char, consignment_file: *mut c_char,
+             transaction_file: *mut c_char) -> Result<(), RequestError> {
     let runtime = Runtime::from_opaque(runtime)?;
-    let data: TransferArgs =
-        serde_json::from_str(ptr_to_string(json)?.as_str()).map_err(|e| format!("{:?}", e))?;
-    info!("{:?}", data);
+
+    let inputs: Vec<OutPoint> =
+        serde_json::from_str(ptr_to_string(inputs)?.as_str()).map_err(|e| format!("{:?}", e))?;
+
+    let allocate: Vec<Outcoins> =
+        serde_json::from_str(ptr_to_string(allocate)?.as_str()).map_err(|e| format!("{:?}", e))?;
+
+    let c_invoice = unsafe { CStr::from_ptr(invoice) };
+    let invoice = Invoice::from_str(c_invoice.to_str()?)?;
+
+    let c_prototype_psbt = unsafe { CStr::from_ptr(prototype_psbt) };
+    let prototype_psbt = c_prototype_psbt.to_str()?.to_string();
+
+    let c_consignment_file = unsafe { CStr::from_ptr(consignment_file) };
+    let consignment_file = c_consignment_file.to_str()?.to_string();
+
+    let c_transaction_file = unsafe { CStr::from_ptr(transaction_file) };
+    let transaction_file = c_transaction_file.to_str()?.to_string();
+
+    info!("TransferArgs {{ inputs: {:?}, allocate: {:?}, invoice: {:?}, prototype_psbt: {:?}, \
+        consignment_file: {:?}, transaction_file: {:?} }}", inputs, allocate, invoice,
+        prototype_psbt, consignment_file, transaction_file);
 
     runtime
         .transfer(
-            data.inputs,
-            data.allocate,
-            data.invoice,
-            data.prototype_psbt,
-            data.consignment_file,
-            data.transaction_file,
+            inputs,
+            allocate,
+            invoice,
+            prototype_psbt,
+            consignment_file,
+            transaction_file,
         )
         .map_err(|e| format!("{:?}", e))
-        .map(|_| ())
+        .map(|_| ())?;
```

Comment:
Very strange construct. Do not know what Alekos meant by it... GitHub does not allow me to edit code here; but simple `?` after `.transfer(..)` will work instead of these three-level nested maps

Reply:
done and added `rgb::i9n::Error` to the `RequestError`